rust/src/delta.rs

```diff
@@ -613,6 +673,127 @@ impl std::fmt::Debug for DeltaTable {
+        // `CommitInfo` is a fairly dynamic data structure so we should work out the data structure approach separately.
+
+        // TODO: calculate isolation level to use when checking for conflicts.
+        // Leaving conflict checking unimplemented for now to get the "single writer" implementation off the ground.
+        // Leaving some commmented code in place as a guidepost for the future.
+
+        // let no_data_changed = actions.iter().all(|a| match a {
+        //     Action::add(x) => !x.dataChange,
+        //     Action::remove(x) => !x.dataChange,
+        //     _ => false,
+        // });
+        // let isolation_level = if no_data_changed {
+        //     IsolationLevel::SnapshotIsolation
+        // } else { 
+        //     IsolationLevel::Serializable
+        // };
+
+        // TODO: create a CommitInfo action and prepend it to actions.
+
+        // try to commit in a loop in case other writers write the next version first
+        let version = self.try_commit_loop(actions).await?;
+
+        self.delta_table.update().await?;
+
+        Ok(version)
+    }
+
+    async fn try_commit_loop(&mut self, actions: &[Action]) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        let max_retries = max_retry_commit_attempts();
+        let mut attempt_number: u32 = 0;
+
+        loop {
+            if attempt_number > max_retries {
+                return Err(DeltaTransactionError::CommitRetriesExceeded);
+            }
+
+            let commit_result = self.try_commit(actions).await;
+
+            if commit_result.is_ok() {
+                return commit_result;
+            }
+
+            attempt_number += 1;
+        }
+    }
+
+    async fn try_commit(&mut self, actions: &[Action]) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        // Serialize all actions that are part of this log entry.
+        let jsons: Vec<String> = actions.iter()
+            .map(|a| {
+                serde_json::to_string(a).unwrap()
+            }).collect();
+        let log_entry = jsons.join("\n");
+
+        // get the next delta table version and the log path where it should be written
+        let attempt_version = self.next_attempt_version().await?;
+        let path = self.delta_table.version_to_log_path(attempt_version);
+
+        // write the log file bytes to storage
+        // rely on storage to fail if the file already exists -
+        // storage must be atomic and fail if file already exists
+        self.delta_table.storage.put_obj(path.as_str(), log_entry.as_bytes()).await?;
+
+        // return the newly committed version
+        Ok(attempt_version)
+    }
+
+    async fn next_attempt_version(&mut self) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        self.delta_table.update().await?;
+        Ok(self.delta_table.version + 1)
+    }
+}
+
+fn max_retry_commit_attempts() -> u32 {
+    std::env::var("DELTA_MAX_RETRY_COMMIT_ATTEMPTS").map_or(10000000, |s| s.parse::<u32>().unwrap())
```

Comment:
It's better to let application decide how they want to read config value at runtime, through config file, env var, etc. In delta.rs, we can make this configurable by making it a config value of the struct itself. This also avoids global side effects and makes it possible to set different max commit attempts for different tables if needed.

Reply:
ah. good thought. perhaps a `DeltaTransactionOptions` to pass to `new` then in case we have other configs to pass in in the future.

Reply:
fixed